lib/ngx/re.lua

```diff
@@ -0,0 +1,74 @@
+-- Copyright (C) Thibault Charbonnier (thibaultcha)
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
+
+
+local re_match_helper = core_regex.re_match_helper
+local tostring = tostring
+local new_tab = base.new_tab
+local sub = string.sub
+
+
+local _M = { version = base.version }
+
+
+function _M.split(subj, regex, opts, ctx, max, res)
+    -- we need to cast this to strings to avoid exceptions when they are
+    -- something else.
+    -- needed because of further calls to string.sub in this function.
+    subj = tostring(subj)
+
+    if not ctx then
+        ctx = new_tab(0, 1)
+    end
+
+    if not ctx.pos then
+        ctx.pos = 1
+    end
+
+    if not res then
+        res = new_tab(0, 0)
+    elseif type(res) ~= "table" then
+        return error("res is not a table", 2)
+    end
+
+    local sub_idx = ctx.pos
+    local res_idx = #res + 1
+    local count = 0
+
+    while true do
+        local from, to, err, caps = re_match_helper(subj, regex, opts, ctx, nil, nil, nil, true)
+        if err then
+            return nil, err
+        end
+
+        if not from then
+            res[res_idx] = sub(subj, sub_idx, #subj)
+            break
+        end
+
+        res[res_idx] = sub(subj, sub_idx, from - 1)
+        count = count + 1
+        res_idx = res_idx + 1
+
+        if max and count == max then
+            break
+        end
+
+        if caps[1] then
+          res[res_idx] = caps[1]
+          res_idx = res_idx + 1
+        end
+
```

Comment:
Better set a trailing `nil` here just in case the user-supplied `res` table has not yet cleared.